src/main/java/com/playmonumenta/networkrelay/RemotePlayerManager.java

```diff
@@ -0,0 +1,347 @@
+		if (remotePlayers == null) {
+			return;
+		}
+
+		mPlugin.getLogger().fine("Unregistering remote shard");
+		Map<String, RemotePlayer> remotePlayerCopy = new ConcurrentSkipListMap<>(remotePlayers);
+		for (Map.Entry<String, RemotePlayer> playerEntry: remotePlayerCopy.entrySet()) {
+			RemotePlayer remotePlayer = playerEntry.getValue();
+			unregisterPlayer(remotePlayer.mUuid);
+		}
+		mRemotePlayerShardMapped.remove(remoteShardName);
+	}
+
+	@EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
+	public void playerJoinEvent(PlayerJoinEvent event) {
+		Player player = event.getPlayer();
+		refreshLocalPlayer(player);
+
+		RemotePlayer remotePlayer = getRemotePlayer(player.getUniqueId());
+		if (remotePlayer != null) {
+			RemotePlayerLoadedEvent remotePLE = new RemotePlayerLoadedEvent(remotePlayer);
+			Bukkit.getServer().getPluginManager().callEvent(remotePLE);
+		}
+	}
+
+	@EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
+	public void playerQuitEvent(PlayerQuitEvent event) {
+		Player player = event.getPlayer();
+		RemotePlayer remotePlayer = new RemotePlayer(player, false);
+		unregisterPlayer(remotePlayer.mUuid);
+		remotePlayer.broadcast();
+	}
+
+	@EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
+	public void networkRelayMessageEvent(NetworkRelayMessageEvent event) {
+		switch(event.getChannel()) {
+			case REMOTE_PLAYER_UPDATE_CHANNEL: {
+				@Nullable JsonObject data = event.getData();
+				if (data == null) {
+					mPlugin.getLogger().severe("Got " + REMOTE_PLAYER_UPDATE_CHANNEL + " channel with null data");
+					return;
+				}
+				remotePlayerChange(data);
+				break;
+			}
+			case REMOTE_PLAYER_REFRESH_CHANNEL: {
+				refreshLocalPlayers();
+				break;
+			}
+			default: {
+				break;
+			}
+		}
+	}
+
+	public static class RemotePlayer {
+		public final UUID mUuid;
+		public final String mName;
+		public final String mShard;
+		public final boolean mIsOnline;
+		private final ConcurrentMap<String, JsonObject> mPluginData;
+
+		public RemotePlayer(Player player, boolean isOnline) {
+			mUuid = player.getUniqueId();
+			mName = player.getName();
+			RemotePlayerManager manager = RemotePlayerManager.getInstance();
+			mShard = (manager != null && manager.getShardName() != null) ? manager.getShardName() : NetworkRelayAPI.getShardName();
+			mIsOnline = isOnline;
+			mPluginData = new ConcurrentHashMap<>();
+		}
+
+		public RemotePlayer(JsonObject remoteData) {
+			mUuid = UUID.fromString(remoteData.get("playerUuid").getAsString());
+			mName = remoteData.get("playerName").getAsString();
+			mIsOnline = remoteData.get("isOnline").getAsBoolean();
+			mShard = remoteData.get("shard").getAsString();
+			mPluginData = serializePluginData(remoteData);
+		}
+
+		@Nullable
+		public JsonObject getPluginData(String pluginId) {
+			if (mPluginData.containsKey(pluginId)) {
+				return mPluginData.get(pluginId);
+			}
+			return null;
```

Comment:
should just return the get() directly, since it will return null if it doesn't exist.